plugins/vppv2/aclplugin/descriptor/acl.go

```diff
@@ -0,0 +1,390 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at:
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package descriptor
+
+import (
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+
+	"github.com/ligato/cn-infra/idxmap"
+	"github.com/ligato/cn-infra/logging"
+	"github.com/ligato/vpp-agent/plugins/kvscheduler/api"
+	"github.com/ligato/vpp-agent/plugins/vppv2/aclplugin/aclidx"
+	"github.com/ligato/vpp-agent/plugins/vppv2/aclplugin/descriptor/adapter"
+	"github.com/ligato/vpp-agent/plugins/vppv2/aclplugin/vppcalls"
+	"github.com/ligato/vpp-agent/plugins/vppv2/ifplugin"
+	ifdescriptor "github.com/ligato/vpp-agent/plugins/vppv2/ifplugin/descriptor"
+	"github.com/ligato/vpp-agent/plugins/vppv2/model/acl"
+)
+
+const (
+	// AclDescriptorName is descriptor name
+	AclDescriptorName = "vpp-acl"
+)
+
+// AclDescriptor is descriptor for ACL
+type AclDescriptor struct {
+	// dependencies
+	log        logging.Logger
+	aclHandler vppcalls.ACLVppAPI
+
+	// runtime
+	ifPlugin ifplugin.API
+}
+
+// NewAclDescriptor is constructor for ACL descriptor
+func NewAclDescriptor(aclHandler vppcalls.ACLVppAPI, ifPlugin ifplugin.API,
+	logger logging.PluginLogger) *AclDescriptor {
+	return &AclDescriptor{
+		log:        logger.NewLogger("acl-descriptor"),
+		ifPlugin:   ifPlugin,
+		aclHandler: aclHandler,
+	}
+}
+
+// GetDescriptor returns descriptor suitable for registration (via adapter) with
+// the KVScheduler.
+func (d *AclDescriptor) GetDescriptor() *adapter.AclDescriptor {
+	return &adapter.AclDescriptor{
+		Name:        AclDescriptorName,
+		NBKeyPrefix: acl.Prefix,
+		KeySelector: func(key string) bool {
+			return strings.HasPrefix(key, acl.Prefix)
+		},
+		ValueTypeName: proto.MessageName((*acl.Acl)(nil)),
+		KeyLabel: func(key string) string {
+			name, _ := acl.ParseNameFromKey(key)
+			return name
+		},
+		ValueComparator: d.EquivalentACLs,
+		WithMetadata:    true,
+		MetadataMapFactory: func() idxmap.NamedMappingRW {
+			return aclidx.NewAclIndex(d.log, "vpp-acl-index")
+		},
+		Add:                d.Add,
+		Delete:             d.Delete,
+		Modify:             d.Modify,
+		ModifyWithRecreate: d.ModifyWithRecreate,
+		IsRetriableFailure: d.IsRetriableFailure,
+		Dependencies:       d.Dependencies,
+		DerivedValues:      d.DerivedValues,
+		Dump:               d.Dump,
+		DumpDependencies:   []string{ifdescriptor.InterfaceDescriptorName},
+	}
+}
+
+// EquivalentACLs compares two ACLs
+func (d *AclDescriptor) EquivalentACLs(key string, oldACL, newACL *acl.Acl) bool {
+	return proto.Equal(oldACL, newACL)
```

Comment:
Here compare only rules with proto.Equal (assuming that interface bindings will be derived) - derived values will be compared by scheduler (also order of ingress/egress interfaces will not matter).